compiler/src/pipeline.rs

```diff
@@ -0,0 +1,489 @@
+            let witness = executor::witgen::WitnessGenerator::new(&pil, &constants, query_callback)
+                .with_external_witness_values(external_witness_values)
+                .generate();
+
+            self.log(&format!("Took {}", start.elapsed().as_secs_f32()));
+            witness
+                .into_iter()
+                .map(|(name, c)| (name.to_string(), c))
+                .collect::<Vec<_>>()
+        });
+        self.artifact = Some(Artifact::GeneratedWitness(GeneratedWitness {
+            pil,
+            constants,
+            witness,
+        }));
+        Ok(())
+    }
+
+    pub fn prove(&mut self, backend: BackendType) -> Result<(), Vec<String>> {
+        self.advance_to(Stage::GeneratedWitness)?;
+        let Artifact::GeneratedWitness(GeneratedWitness {
+            pil: analyzed,
+            constants,
+            witness,
+        }) = std::mem::take(&mut self.artifact).unwrap()
+        else {
+            panic!()
+        };
+        let factory = backend.factory::<T>();
+        let backend = factory.create(analyzed.degree());
+
+        // Even if we don't have all constants and witnesses, some backends will
+        // still output the constraint serialization.
+        let (proof, constraints_serialization) = backend.prove(
+            &analyzed,
+            &constants,
+            witness.as_deref().unwrap_or_default(),
+            None,
+        );
+
+        if let Some(output_dir) = &self.output_dir {
+            write_constants_to_fs(&constants, output_dir);
+            if let Some(witness) = &witness {
+                write_commits_to_fs(witness, output_dir);
+            }
+            if let Some(constraints_serialization) = &constraints_serialization {
+                write_constraints_to_fs(constraints_serialization, output_dir);
+            }
+        };
+
+        self.artifact = Some(Artifact::Proof(CompilationResult {
+            constants,
+            witness,
+            proof,
+            constraints_serialization,
+        }));
+        Ok(())
+    }
+
+    pub fn analyzed_asm(mut self) -> Result<AnalysisASMFile<T>, Vec<String>> {
+        self.advance_to(Stage::AnalyzedAsm)?;
+        let Artifact::AnalyzedAsm(analyzed_asm) = self.artifact.unwrap() else {
+            panic!()
+        };
+        Ok(analyzed_asm)
+    }
+
+    pub fn analyzed_asm_ref(&mut self) -> Result<&AnalysisASMFile<T>, Vec<String>> {
+        self.advance_to(Stage::AnalyzedAsm)?;
+        match self.artifact.as_ref().unwrap() {
+            Artifact::AnalyzedAsm(analyzed_asm) => Ok(analyzed_asm),
+            _ => panic!(),
+        }
+    }
+
+    pub fn optimized_pil_with_constants(mut self) -> Result<PilWithConstants<T>, Vec<String>> {
+        self.advance_to(Stage::PilWithConstants)?;
+        let Artifact::PilWithConstants(pil_with_constants) = self.artifact.unwrap() else {
+            panic!()
+        };
+        Ok(pil_with_constants)
+    }
+
```

Comment:
Don't we also need API entries that return the optimized PIL without the constants?